src/iop/blurs.c

```diff
@@ -0,0 +1,855 @@
+
+  // lens blur params
+  int blades;              // $MIN: 3 $MAX: 11 $DEFAULT: 5 $DESCRIPTION: "diaphragm blades"
+  float concavity;         // $MIN: 1. $MAX: 9.  $DEFAULT: 1. $DESCRIPTION: "concavity"
+  float linearity;         // $MIN: 0. $MAX: 1.  $DEFAULT: 1. $DESCRIPTION: "linearity"
+  float rotation;          // $MIN: -1.57 $MAX: 1.57 $DEFAULT: 0. $DESCRIPTION: "rotation"
+
+  // motion blur params
+  float angle;             // $MIN: -3.14 $MAX: 3.14 $DEFAULT: 0. $DESCRIPTION: "direction"
+  float curvature;         // $MIN: -2.   $MAX: 2.   $DEFAULT: 0. $DESCRIPTION: "curvature"
+  float offset;            // $MIN: -1.   $MAX: 1.   $DEFAULT: 0  $DESCRIPTION: "offset"
+
+} dt_iop_blurs_params_t;
+
+
+typedef struct dt_iop_blurs_gui_data_t
+{
+  GtkWidget *type, *radius, *blades, *concavity, *linearity, *rotation, *angle, *curvature, *offset;
+  GtkDrawingArea *area;
+  unsigned char *img;
+  int img_cached;
+  float img_width;
+} dt_iop_blurs_gui_data_t;
+
+
+typedef struct dt_iop_blurs_global_data_t
+{
+  int kernel_blurs_convolve;
+} dt_iop_blurs_global_data_t;
+
+
+const char *name()
+{
+  return _("blurs");
+}
+
+const char *aliases()
+{
+  return _("blur|lens|motion");
+}
+
+const char *description(struct dt_iop_module_t *self)
+{
+  return dt_iop_set_description(self,
+                                _("simulate physically-accurate lens and motion blurs."),
+                                _("creative"), _("linear, RGB, scene-referred"), _("linear, RGB"),
+                                _("linear, RGB, scene-referred"));
+}
+
+int flags()
+{
+  return IOP_FLAGS_INCLUDE_IN_STYLES | IOP_FLAGS_SUPPORTS_BLENDING;
+}
+
+
+int default_group()
+{
+  return IOP_GROUP_EFFECTS | IOP_GROUP_EFFECT;
+}
+
+
+int default_colorspace(dt_iop_module_t *self, dt_dev_pixelpipe_t *pipe, dt_dev_pixelpipe_iop_t *piece)
+{
+  return iop_cs_rgb;
+}
+
+
+void commit_params(dt_iop_module_t *self, dt_iop_params_t *p1, dt_dev_pixelpipe_t *pipe, dt_dev_pixelpipe_iop_t *piece)
+{
+  memcpy(piece->data, p1, self->params_size);
+}
+
+// B spline filter
+#define FSIZE 5
+
+inline static void blur_2D_Bspline(const float *const restrict in, float *const restrict out,
+                                   const size_t width, const size_t height)
+{
+#ifdef _OPENMP
+#pragma omp parallel for default(none) dt_omp_firstprivate(width, height, in, out) \
+    schedule(simd: static)    \
+    collapse(2)
+#endif
+  for(size_t i = 0; i < height; i++)
+  {
+    for(size_t j = 0; j < width; j++)
+    {
+      const size_t index = (i * width + j);
+      float acc = 0.f;
+
+      for(size_t ii = 0; ii < FSIZE; ++ii)
+        for(size_t jj = 0; jj < FSIZE; ++jj)
+        {
+          const size_t row = CLAMP((int)i + (int)(ii - (FSIZE - 1) / 2), (int)0, (int)height - 1);
+          const size_t col = CLAMP((int)j + (int)(jj - (FSIZE - 1) / 2), (int)0, (int)width - 1);
+          const size_t k_index = (row * width + col);
+
+          const float DT_ALIGNED_ARRAY filter[FSIZE]
```

Comment:
I'm not sure how the optimizer can handle this but I would certainly put filter const array outside of the loops.

Reply:
Given discussion with ralfbrown on another PR, just as static keyword will suffice here.

Reply:
I don't understand the static. Isn't that supposed to be reserved for variable that may be changed along the life of the program ?

Reply:
static const will place the variable into a data segement which is appropriate here.

Reply:
fixed